api/src/db/db.service.spec.ts

```diff
@@ -164,6 +164,59 @@ describe("DbService", () => {
         expect(docCount).toBe(8);
     });
 
+    it("can detect that a document is exactly the same as the document in the database", async () => {
+        const res: any = await service.upsertDoc({
+            _id: "group-private-content",
+            type: "group",
+            updatedTimeUtc: 3,
+            name: "Private Content",
+            acl: [
+                {
+                    type: "post",
+                    groupId: "group-private-users",
+                    permission: ["view"],
+                },
+                {
+                    type: "tag",
+                    groupId: "group-private-users",
+                    permission: ["view"],
+                },
+                {
+                    type: "post",
+                    groupId: "group-private-editors",
+                    permission: ["view", "edit", "translate", "publish"],
+                },
+                {
+                    type: "tag",
+                    groupId: "group-private-editors",
+                    permission: ["view", "translate", "assign"],
+                },
+                {
+                    type: "group",
+                    groupId: "group-private-editors",
+                    permission: ["view", "assign"],
+                },
+            ],
+        });
+
+        expect(res).toBe("passed document equal to existing database document");
+    });
+
+    it("can handle simultaneous updates to a single document", async () => {
+        const pList = new Array<Promise<any>>();
+
+        for (let index = 0; index < 50; index++) {
+            pList.push(service.upsertDoc({ _id: "simultaneousTest", testVal: index }));
+        }
+
+        let res: boolean = false;
+        await Promise.all(pList);
+
+        // if we got past this point without an exception, the test was successful
+        res = true;
+        expect(res).toBe(true);
```

Comment:
Suggestion: get the document here and assert the index is what we expect. Maybe the method silently fails and it didn't actually work.

Reply:
Updating documents with the wrong revision ID throws an exception. If the Promise.all() runs without exceptions the logic handles the exceptions correctly by re-trying to get the latest revision ID and updating the document again. Looking at the index value will not work as we do not know which simultaneous update is actually ending up being the final update. (The whole idea is not to have sequential updates but to simulate a real-world scenario where two users update a document at exactly the same time.)

Changed the number of simultaneous updates to 50 which should give a decent amount of real duplicate updates to test the logic (hits the exception handling around 10 times for the 50 document updates in my development environment).

Reply:
Ah ok, and then we rely on our changelog to view the actual changes in the CMS in case someone ends up with an unexpected version?